src/cpp/statistics/rtps/messages/RTPSStatisticsMessages.hpp

```diff
@@ -0,0 +1,195 @@
+}
+
+/**
+ * @brief Read a statistics submessage from a message.
+ * @param [in,out] msg Message from where to extract the submessage.
+ * @param [out] data Data read from the statistics submessage.
+ * @pre msg->pos should point to the beginning of the statistics submessage payload
+ * @post msg->length will be decremented by @c statistics_submessage_length (i.e. the
+ *       submessage will be consumed, and it won't be available anymore)
+ */
+inline void read_statistics_submessage(
+        eprosima::fastrtps::rtps::CDRMessage_t* msg,
+        StatisticsSubmessageData& data)
+{
+    static_cast<void>(msg);
+    static_cast<void>(data);
+
+#ifdef FASTDDS_STATISTICS
+    // Should be exactly at the end
+    assert(msg->pos + statistics_submessage_data_length == msg->length);
+
+    // Read all fields
+    using namespace eprosima::fastrtps::rtps;
+    CDRMessage::readInt32(msg, &data.ts.seconds);
+    CDRMessage::readUInt32(msg, &data.ts.fraction);
+    CDRMessage::readUInt64(msg, &data.seq.sequence);
+    CDRMessage::readUInt64(msg, &data.seq.bytes);
+    CDRMessage::readUInt16(msg, &data.seq.bytes_high);
+
+    // Consume submessage
+    msg->length -= statistics_submessage_length;
+    msg->pos = msg->length;
+#endif // FASTDDS_STATISTICS
+}
+
+#ifdef FASTDDS_STATISTICS
+inline uint32_t get_statistics_message_pos(
+        const eprosima::fastrtps::rtps::octet* send_buffer,
+        uint32_t send_buffer_size)
+{
+    // Message should contain RTPS header and statistic submessage
+    assert(statistics_submessage_length + RTPSMESSAGE_HEADER_SIZE <= send_buffer_size);
+
+    // The last submessage should be the statistics submessage
+    uint32_t statistics_pos = send_buffer_size - statistics_submessage_length;
+    assert(FASTDDS_STATISTICS_NETWORK_SUBMESSAGE == send_buffer[statistics_pos]);
+
+    return statistics_pos;
+}
+
+#endif // FASTDDS_STATISTICS
+
+inline void set_statistics_submessage_from_transport(
+        const eprosima::fastrtps::rtps::octet* send_buffer,
+        uint32_t send_buffer_size,
+        StatisticsSubmessageData::Sequence& sequence)
+{
+    static_cast<void>(send_buffer);
+    static_cast<void>(send_buffer_size);
+    static_cast<void>(sequence);
+
+#ifdef FASTDDS_STATISTICS
+    using namespace eprosima::fastrtps::rtps;
+
+    uint32_t statistics_pos = get_statistics_message_pos(send_buffer, send_buffer_size);
+
+    // Accumulate bytes on sequence
+    sequence.add_message(send_buffer_size);
+
+    // Skip the submessage header
+    statistics_pos += RTPSMESSAGE_SUBMESSAGEHEADER_SIZE;
+
+    // Set current timestamp and sequence
+    auto submessage = (StatisticsSubmessageData*)(&send_buffer[statistics_pos]);
+    Time_t ts;
+    Time_t::now(ts);
+
+    submessage->ts.seconds = ts.seconds();
+    submessage->ts.fraction = ts.fraction();
+    submessage->seq.sequence = sequence.sequence;
+    submessage->seq.bytes = sequence.bytes;
+    submessage->seq.bytes_high = sequence.bytes_high;
```

Comment:
submessage->seq = sequence;

Reply:
Turns out the compiler optimization (msvc) copies the padding too. This padding is non-zeroed in stack objects and breaks RTPS standard directive.